File: 2022/day3/day3.py
datafile = open("day3/input.txt", "r")
# datafile = open("day3/test.txt", "r")

def getPriority(x):
    if x.islower():
       return ord(x)-96
    else:
       return ord(x)-38

sacks = [x.strip() for x in datafile]
priority = 0
for sack in sacks:
    splitPoint = int(len(sack) / 2)
    commonItem = list(set(sack[0:splitPoint]).intersection(set(sack[splitPoint:])))[0]    
    priority += getPriority(commonItem)
print("Answer Part 1: ", priority)

priority = 0
sacksThree =  [sacks[i:i + 3] for i in range(0, len(sacks), 3)]
for sack in sacksThree:
    commonItem = list(set(sack[0]).intersection(set(sack[1])).intersection(set(sack[2])))[0]
    priority += getPriority(commonItem)
print("Answer Part 2: ", priority)

File: 2022/day3/day3.hs
module Main where

import Data.Char ( ord, isLower )
import Data.List ( intersect, nub )
import Data.List.Split ( chunksOf )

main :: IO ()
main = do
        dataFile <- readFile "day3/input.txt"
        let sacks1 = (map splitInHalf . lines) dataFile
        let commonItem1 = map (head . nub . uncurry intersect) sacks1

        putStr "Part 1 answer: " 
        print $ (sum . map getPriority) commonItem1

        let sacks2 = (chunksOf 3 . lines) dataFile
        let commonItem2 = map (head . nub . intersect3 ) sacks2

        putStr "Part 2 answer: " 
        print $ (sum . map getPriority) commonItem2

getPriority :: Char -> Int
getPriority x =
    case isLower x of 
        True -> ord x - 96
        otherwise -> ord x - 38

splitInHalf :: [a] -> ([a], [a])
splitInHalf x = splitAt (length x `div` 2) x

intersect3 :: Eq a => [[a]] -> [a]
intersect3 [x,y,z] = intersect (intersect x y) z 